extensions/omni_vfs/migrate/4_table_fs.sql

```diff
@@ -0,0 +1,155 @@
+    /*  external storage optimizes fetching parts of data
+        at the expense of increased storage due to disabled compression
+    */
+    data bytea  storage external not null
+);
+
+create type table_fs as
+(
+    id integer
+);
+
+create function table_fs(filesystem_name text) returns omni_vfs.table_fs
+    language plpgsql
+as
+$$
+declare
+    result_id integer;
+begin
+    select
+        id
+    from
+        omni_vfs.table_fs_filesystems
+    where
+        table_fs_filesystems.name = filesystem_name
+    into result_id;
+    if not found then
+        insert
+        into
+            omni_vfs.table_fs_filesystems (name)
+        values (filesystem_name)
+        returning id into result_id;
+    end if;
+    return row (result_id);
+end
+$$;
+
+create function list(fs table_fs, path text, fail_unpermitted boolean default true) returns setof omni_vfs_types_v1.file
+    stable
+    language sql
+as
+$$
+with
+    filepath(path) as (
+        select
+            case
+                when omni_vfs.canonicalize_path(path, absolute => true) = '/' then ''
+                else omni_vfs.canonicalize_path(path, absolute => true)
+            end
+    )
+select
+    distinct case
+        -- filename exactly matches filepath.path
+        when filename = filepath.path then row(omni_vfs.basename(filename), kind)::omni_vfs_types_v1.file
+        -- filepath.path is prefix of filename
+        else (
+                case
+                    -- only one '/' left, eg. /file
+                    when split_part(substr(filename, length(filepath.path)+1), '/', 3) = '' then
+                        row(split_part(substr(filename, length(filepath.path)+1), '/', 2), kind)::omni_vfs_types_v1.file
+                    -- more than one '/', eg. /dir/file
+                    else row(split_part(substr(filename, length(filepath.path)+1), '/', 2), 'dir')::omni_vfs_types_v1.file
+                end
+        )
+    end
+from omni_vfs.table_fs_files, filepath
+where filesystem_id = fs.id and
+(filename like (filepath.path || '/%') or
+filename = filepath.path)
+$$;
+
+create function file_info(fs table_fs, path text) returns omni_vfs_types_v1.file_info
+    stable
+    language sql
+as
+$$
+with
+    filepath(path) as (select omni_vfs.canonicalize_path(path, absolute => true))
+    select
+        coalesce(length(data), 0) as size,
+        created_at,
+        accessed_at,
+        modified_at,
+        kind
+    from omni_vfs.table_fs_files cross join filepath
+    left join omni_vfs.table_fs_file_data on id = file
+    where filesystem_id = fs.id and filename = filepath.path
+$$;
+
+
+create function read(fs table_fs, path text, file_offset bigint default 0,
+                     chunk_size int default null) returns bytea
+    stable
+    language sql
+as
+$$
+with
+    filepath(path) as (select omni_vfs.canonicalize_path(path, absolute => true))
+    select
+        case
+            when chunk_size is null then substr(data, file_offset::int)
```

Comment:
postgres inbuilt `substr` expects offset parameter to be of type `int` so there is a potential concern of truncation of value due to downcasting from `bigint` to `int`